src/worker/ingestion/action-matcher.ts

```diff
@@ -0,0 +1,498 @@
+    }
+
+    /**
+     * Sublevel 3 or 5 of action matching.
+     */
+    public checkElementsAgainstSelector(elements: Element[], selector: string, escapeSlashes = true): boolean {
+        const parts: SelectorPart[] = []
+        // Sometimes people manually add *, just remove them as they don't do anything
+        selector = selector
+            .replace(/> \* > /g, '')
+            .replace(/> \*/g, '')
+            .trim()
+        const tags = selector.split(' ')
+        // Detecting selector parts
+        for (let partIndex = 0; partIndex < tags.length; partIndex++) {
+            const tag = tags[partIndex]
+            if (tag === '>' || tag === '') {
+                continue
+            }
+            const directDescendant = partIndex > 0 && tags[partIndex - 1] === '>'
+            const part = new SelectorPart(tag, directDescendant, escapeSlashes)
+            part.uniqueOrder = parts.filter((p) => equal(p.requirements, part.requirements)).length
+            parts.push(part)
+        }
+        // Matching elements against selector parts
+        // Initial base element is the imaginary parent of the outermost known element
+        let baseElementIndex = elements.length
+        let wasPartMatched = true
+        let partIndex = 0
+        while (partIndex < parts.length) {
+            if (baseElementIndex <= 0) {
+                // At least one part wasn't matched yet, but at the same time there are no more matchable elements left!
+                return false
+            }
+            const part = parts[partIndex]
+            wasPartMatched = false
+            for (let depthDiff = 1; baseElementIndex - depthDiff >= 0; depthDiff++) {
+                // Subtracting depthDiff as elements are reversed, meaning outer elements have higher indexes
+                const currentElementIndex = baseElementIndex - depthDiff
+                if (
+                    this.checkElementAgainstSelectorPartRequirements(elements[currentElementIndex], part.requirements)
+                ) {
+                    baseElementIndex = currentElementIndex
+                    wasPartMatched = true
+                    break
+                } else if (part.directDescendant) {
+                    break
+                }
+            }
+            if (wasPartMatched) {
+                partIndex++
+            } else {
+                if (part.directDescendant) {
+                    // For a direct descendant we need to check the parent parent again
+                    partIndex--
+                } else {
+                    // Otherwise we just move the base down
+                    baseElementIndex--
+                }
+            }
+        }
+        return wasPartMatched
+    }
+
+    private checkElementAgainstSelectorPartRequirements(element: Element, requirements: Partial<Element>): boolean {
+        if (requirements['text'] && element.text !== requirements.text) {
```

Comment:
nit: in some places this check is done with `a.b` and others `a['b']`

Reply:
Yeah, leftover from some code ported from Python where dicts require brackets.